source/opt/flatten_array_matrix_stage_var.h

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_OPT_FLATTEN_ARRAY_MATRIX_STAGE_VAR_H_
+#define SOURCE_OPT_FLATTEN_ARRAY_MATRIX_STAGE_VAR_H_
+
+#include <unordered_set>
+
+#include "source/opt/pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// A struct for the information of a stage variable's location, component,
+// extra arrayness, and whether it is an input or output stage variable. Note
+// that stage variables of the tessellation shaders have the extra arrayness as
+// the first dimension of the array.
+//
+// For example, when we generate the following HLSL code to SPIR-V
+//
+//  struct HS_OUTPUT {
+//    float3 pos[5] : POSITION;
+//    ...
+//  };
+//  ...
+//  [patchconstantfunc("main_hs_patch")]
+//  [outputcontrolpoints(3)]
+//  HS_OUTPUT main_hs(...) { ... }
+//  ...
+//
+//  HS_PATCH_OUTPUT main_hs_patch(const OutputPatch<HS_OUTPUT, 3> patch) {
+//  ...
+//  }
+//
+// because of `OutputPatch<HS_OUTPUT, 3> patch`, we add the extra arrayness 3
+// for `float3 pos[5]`. In SPIR-V, its type will be
+// `%_ptr_Output__arr__arr_v3float_uint_5_uint_3`.
+struct StageVariableInfo {
+  uint32_t location;
+  uint32_t component;
+  uint32_t extra_arrayness;
+  bool is_input_var;
+
+  bool operator==(const StageVariableInfo& another) const {
+    return another.location == location && another.component == component &&
+           another.is_input_var == is_input_var;
+  }
+};
+
+// See optimizer.hpp for documentation.
+//
+// Note that the current implementation of this pass covers only store, load,
+// access chain instructions for the stage variables. Supporting other types of
+// instructions is a future work.
+class FlattenArrayMatrixStageVariable : public Pass {
+ public:
+  // Hashing functor for StageVariableInfo.
+  struct StageVariableInfoHash {
+    size_t operator()(const StageVariableInfo& info) const {
+      return std::hash<uint32_t>()(info.location) ^
+             std::hash<uint32_t>()(info.component) ^
+             std::hash<uint32_t>()(static_cast<uint32_t>(info.is_input_var));
+    }
+  };
+
+  using SetOfStageVariableLocationInfo =
+      std::unordered_set<StageVariableInfo, StageVariableInfoHash>;
+
+  explicit FlattenArrayMatrixStageVariable(
+      const std::vector<StageVariableInfo>& stage_variable_info)
+      : stage_variable_info_(stage_variable_info.begin(),
+                             stage_variable_info.end()) {}
```

Comment:
I am not sure why you need the "stage variables" to be passed into the pass?  Are there some input/output variables that should not be replaced by scalars or vectors?  If all of them should be replaced, then this information is redundant.  We can just find all of the variables in the module.  If the issue is when there are multiple entry points (because tesselation shaders are handled differently), we could probably find another way to deal with that.